### ISSUE TITLE
RDF4J log is full of "Deregistered expired transactions" warnings - potential source of issues?

### ISSUE THREAD

**Issue by tomas-knap** [eclipse/rdf4j#1559]
We are using HTTP Repository (so native via http) and In the the rdf4j-server, there are lots of warnings like:
```
[WARN ] 2019-09-25 00:37:56,788 [repositories/uv31] deregistered expired transaction 66feb7f0-6786-4478-845b-7ee64107e80b
[WARN ] 2019-09-25 00:37:56,788 [repositories/uv31] deregistered expired transaction e8b9e3ff-f01d-4552-8463-349ab81ae9dd
[WARN ] 2019-09-25 00:37:56,790 [repositories/uv31] deregistered expired transaction 675f1a0d-6f93-4a9e-9319-4bff741d3838
[WARN ] 2019-09-25 00:37:56,790 [repositories/uv31] deregistered expired transaction 7200cf0b-9adb-4210-a3b2-c7ffb8eb0074
[WARN ] 2019-09-25 00:37:56,790 [repositories/uv31] deregistered expired transaction abdf3a24-530a-4630-8310-70aba8f0923a
[WARN ] 2019-09-25 00:37:56,790 [repositories/uv31] deregistered expired transaction db1b359e-ce90-4e9d-88c5-2949e7ac3768
[WARN ] 2019-09-25 00:37:56,790 [repositories/uv31] deregistered expired transaction a194f5ca-4653-40bf-8dd3-df98080e2e80
[WARN ] 2019-09-25 00:37:56,790 [repositories/uv31] deregistered expired transaction f08915be-0e82-429e-86e9-d19bf0d28adb
[WARN ] 2019-09-25 00:37:56,790 [repositories/uv31] deregistered expired transaction 250f75f7-f6c6-4de5-ae02-4cb9ade8d913
```
Is it something extraordinary we should be worried about - e.g. we are not properly closing transactions (we basically finish the transaction with either commit() or rollback(), which should be sufficient))? Or deregistering expired transactions is something which happens normally for all the transactions when they are commited/rollbacked and the WARN level is not the most appropriate here? 

The code which is responsible for logging that message in RDF4J: 
```
	if (RemovalCause.EXPIRED.equals(notification.getCause())) {
					final UUID transactionId = notification.getKey();
					final Transaction entry = notification.getValue();
					synchronized (primaryCache) {
						if (!entry.hasActiveOperations()) {
							// no operation active, we can decommission this entry
							primaryCache.invalidate(transactionId);
							logger.warn("deregistered expired transaction {}", transactionId);
						}
						else {
							// operation still active. Reinsert in secondary cache.
							secondaryCache.put(transactionId, entry);
						}
					}
				}
```

We are using rdf4j 2.3.3

**Comment by tomas-knap** [comment]
Could be related with https://github.com/eclipse/rdf4j/issues/725

**Comment by hmottestad** [comment]
I don’t know anything about this aspect of RDF4J so I can only answer how transactions should be handled.

A transaction should either finish with a rollback or a commit. All read/select operations on a connection need to be closed. All connections need to be closed.

**Comment by tomas-knap** [comment]
@hmottestad Thanks for reply, so what is the typical flow which can lead to "deregistered expired transaction X" messages? (we are using httpRepo and properly closing transactions)

**Comment by hmottestad** [comment]
I can't really help much with that. I don't know much about the httpRepo code and never really used it for any of my projects.

**Comment by hmottestad** [comment]
@jeenbroekstra would know if anyone knows.

**Comment by jeenbroekstra** [comment]
"deregistered expired transaction X" happens when the server-side transaction registry (which uses a cache with a timeout mechanism) finds "orphaned" transactions: transactions that have not been closed but which have not been active for at least N seconds (by default, 60). So normally I'd say that the only way this can happen (other than opening a transaction and then not doing anything with it for 60 seconds) is if you are not ending your transactions properly on the client side - and that's why this is logged as a warning.

A best practice is to make sure you close the connection when you're done with it, too, as that will close the transaction. Ideally, use a try-with-resources so that the connection gets closed even if an exception is thrown.

**Comment by jeenbroekstra** [comment]
Btw since you're using 2.3.3, it's possible that you are experiencing a bug that has since been fixed though. I'm fairly sure I remember we did a few fixes and improvements in the transaction registry in the 2.4.x releases (I'll try and look up the specific bugs later). Can you try upgrading to a newer version of RDF4J and see if the problem persists?

**Comment by tomas-knap** [comment]
Thanks @jeenbroekstra, if for some reason transaction is not closed properly (e.g. there is an exception during commit which causes it is at the end not committed/rolled back) and it is dangling there - will it be marked as orphaned after 60s (whatever the timeout is), or will it be there forever (until the client stops), because of that transaction PING mechanism, which would keep the transaction alive forever (until the client stops)?

Regarding fixes in 2.4.X, I found mainly that one, which I will try:  
- https://github.com/eclipse/rdf4j/issues/1260

**Comment by jeenbroekstra** [comment]
@tomas-knap the RDF4JProtocolSession object keeps a transaction alive from the client-side by using pings - this prevents the server from discarding the transaction prematurely. It does this until the transaction gets closed (by rollback or commit).

The life-time of an RDF4JProtocolSession is determined by the lifetime of the HTTPConnection that created it. As soon as the connection gets closed, the session gets closed - which should stop any further pings.

The only scenario I can think of where it would keep sending pings indefinitely would be if you have a HTTPRepositoryConnection on which a transaction was started but never committed/rolled back, and which was also never closed. Even if it were garbage-collected, it might be possible that the underlying session object would remain active, and continue sending pings. 

In short: both this issue and the related issue #1408 seem to me to be caused by not properly closing connections at the client-side. At least that's the only path I can see to get the described behaviour. However, if you can show a code example that actually demonstrates these messages occurring in a situation where all connections are properly closed, we can dig a bit deeper.

**Comment by tomas-knap** [comment]
@jeenbroekstra Thanks, but as I tried to explain now in https://github.com/eclipse/rdf4j/issues/1408, in that case it is NOT possible to close connection with a broken transaction as connection.close() throws exception - so there are pings running forever and connection dangling there.
